### PR TITLE
[MERGE] web,*: update code base to jQuery 3.*

### PR DESCRIPTION
**Pull request by VincentSchippefilt**
The goal of this task is to update jQuery from version 1.11 to 3.x.
Task-id : 1896658

As we are skipping 2 major versions, there are some breaking changes in the jquery library described here: https://jquery.com/upgrade-guide/3.0/

Main changes are:
-------------------------
1. Update the libraries jQuery, jQueryUI, QUnit and select2 to their latest version.
   The link to generate the build of jQueryUI with the used components: [jquery UI build](https://jqueryui.com/download/#!version=1.12.1&components=111110010101111110100010110000011010000000000000)
2. Use native promises ([a complete explanation about promises here](https://github.com/getify/You-Dont-Know-JS/blob/master/async%20%26%20performance/ch3.md)) everywhere possible:
In order to [be A+ compliant](https://promisesaplus.com/), the jq3 deferred uses `setTimeout(..., 0)`
for every then and catch, which takes between 4 and 10 ms by call.

    This would be way too slow to use in Odoo.
    To prevent this, we have used the native Promises everywhere possible.
    Be it with jq deferred or native promises, all promises are now asynchronous,
    this example will ALWAYS result in printing first "just after blabla" then "blabla"
    because the `then` function will always be asynchronous (opposite to
    what jquery used to do)
    ```javascript
    function () {
        Promise.resolve("blabla").then(console.log);
        console.log("just after blabla");
    }

    // console:
    // just after blabla
    // blabla
    ```

3. On native promises, you should see the catch like a try/catch in
javascript, meaning that it will catch all errors including coding errors:
    For this we have introduced the concept of `guardedCatch`, which will be executed
    on promise rejection unless the promise is rejected because there is
    a code error or runtime error (calls on undefined, etc).

4. Native Promises do not have `done`, `fail` and `always` methods nor does it have the `state` method

    - `fail` can be replaced with `guardedCatch`
    - `return $.when(x).done(function() ...)` can be replaced with

    ```javascript
    var x = Promise.resolve(x); // we get the x promise in a local variable, without then
    x.then(function() ...); // we add a new handler to x. This operation returns a new promise we don't care about
    return x; // we return the x promise that will be completed without waiting for the then
    ```

    - `$.when().done(function ...)` without the return can be replaced by `then(function ....)`, because if it is not returned, it doesn't matter what is chained or not `$.when().always(function() ...)` can be replaced with
    ```javascript
    var always = function() ...
    Promise.resolve().then(always).guardedCatch(always);
    ```

    - It is not possible to inspect the state of a promise, except in debug tools of chrome and firefox. Any code that relies on `$.Deferred.state()` should be rewritten to use flags in `then` and `catch` handlers. Remember that it is not possible to execute the handlers in a synchronous manner, so when a promise is created, its `then` will never be called on the same execution stack. At best it will be called on the next microtask tick (before the next setTimeout).

5. We have introduced the concept of guardedCatch, to copy the behavior of the old fail:
   The guarded catch will only execute the handler in parameter if the rejection reason is not an `Error`
    ```javascript
    var failingPromise = Promise.reject({message:"I am the rejection reason"});
    failingPromise.guardedCatch(function(reason) {console.log(reason);}); // is called, then returns a rejected promise
    failingPromise.catch(function(reason) {console.log(reason);}); // is called
    ```
    will write the reason "I am the rejection reason" to the console twice: once because the `guardedCatch` will execute (and return a rejected `Promise` with the same reason it was called), a second time because the `catch` will catch the rejected reason returned by the `guardedCatch`
    ```javascript
    var promiseWithError = Promise.resolve().then(function() {
        a.b = 0;  // <-- the variable a does not exist, so a.b is an error
    });
    promiseWithError.guardedCatch(function(reason) {console.log(reason);}); // will not be called
    promiseWithError.catch(function(reason) {console.log(reason);}); // will be called
    ```
    contrary to how deferred used to work, here the catch will be called because it behaves like a `try... catch...` and will catch the coding error like if it was a business error. The `guardedCatch` however does not call the handler if the rejection reason is an error.

    In all Odoo code, we never reject a promise with an error and always use objects to defined the rejected reason, so it is a good idea to use `guardedCatch` everywhere.

6. Native Promises can have only one parameter passed to a `then` function or a `catch` function:
    `xxx.then(function(param1, param2) {...}); // param2 will never have a value`

    - we can replace `$.when(prom1, prom2).then(function(result1, result2) { ... });` by
    ```javascript
    Promise.all([prom1, prom2]).then(function(results) {
        var result1 = results[0];
        var result2 = results[1];
        ...
    });
    ```

7. We are now using ES7 code in the unit tests (async await).

Co-authored-by: Aaron Bohy <aab@odoo.com>
Co-authored-by: Christophe Matthieu <chm@odoo.com>
Co-authored-by: Mathieu Duckerts-Antoine <dam@odoo.com>
Co-authored-by: David Monjoie <dmo@odoo.com>
Co-authored-by: Martin Geubelle <mge@odoo.com>
Co-authored-by: svs-odoo <svs@odoo.com>
Co-authored-by: Vincent Schippefilt <vsc@odoo.com>

### CODE REVIEW

File: addons/web/static/src/js/chrome/action_manager.js [hunk 1]
```diff
@@ -488,7 +488,7 @@ var ActionManager = Widget.extend({
         });
         this.controllers[controllerID] = controller;
         action.controllerID = controllerID;
-        return this._executeAction(action, options).done(function () {
```
Comment (aab-odoo) [inline addons/web/static/src/js/chrome/action_manager.js]: be careful with this

File: addons/web/static/tests/core/concurrency_tests.js [hunk 2]
```diff
@@ -116,99 +147,127 @@ QUnit.module('core', {}, function () {
...
 
         var dp = new concurrency.DropPrevious();
 
-        var def1 = $.Deferred();
-        var def2 = $.Deferred();
+        var prom1 = makeTestPromise(assert, 'prom1');
+        var prom2 = makeTestPromise(assert, 'prom2');
 
-        var r1 = dp.add(def1);
+        var dp1 = dp.add(prom1).then(() => assert.catch('should not go here'))
```
Comment (aab-odoo) [inline addons/web/static/tests/core/concurrency_tests.js]: .step

File: addons/mail/static/src/js/chatter.js [hunk 3]
```diff
@@ -93,10 +93,10 @@ var Chatter = Widget.extend({
         }));
         // start and append the widgets
         var fieldDefs = _.invoke(this.fields, 'appendTo', $('<div>'));
-        var def = this._dp.add($.when.apply($, fieldDefs));
-        this._render(def).then(this._updateMentionSuggestions.bind(this));
+        var def = this._dp.add(Promise.all(fieldDefs));
+        var renderPromise = this._render(def).then(this._updateMentionSuggestions.bind(this));
 
-        return this._super.apply(this, arguments);
+        return Promise.all([this._super.apply(this, arguments), renderPromise]);
```
Comment (aab-odoo) [inline addons/mail/static/src/js/chatter.js]: you change the behavior here

File: addons/web/static/src/js/chrome/action_manager.js (hunk 1 above)
Comment (VincentSchippefilt) [reply addons/web/static/src/js/chrome/action_manager.js]: yes, we have updated, thanks

File: addons/web/static/tests/core/concurrency_tests.js (hunk 2 above)
Comment (VincentSchippefilt) [reply addons/web/static/tests/core/concurrency_tests.js]: fixed

File: addons/mail/static/src/js/chatter.js (hunk 3 above)
Comment (VincentSchippefilt) [reply addons/mail/static/src/js/chatter.js]: indeed. We will re-optimize if needed.

**Comment by aab-odoo** [comment]
robodoo rebase-merge

**Comment by aab-odoo** [comment]
robodoo r+

**Comment by aab-odoo** [comment]
robodoo merge

**Comment by aab-odoo** [comment]
robodoo r+

**Comment by aab-odoo** [comment]
robodoo r-

**Comment by aab-odoo** [comment]
robodoo r+

**Comment by dmo-odoo** [comment]
@aab-odoo @VincentSchippefilt ❤️

**Comment by xmo-odoo** [comment]
> * `$.when().always(function() ...)` can be replaced with

In most cases we probably want `Promise#finally` though, especially as it doesn't "chain" the promise unless the callback blows up (that is unless `always` throws an exception `Promise.resolve(1).finally(always)` will return the same as `Promise.resolve(1)` and `Promise.reject(1).finally(always)` will return the same as `Promise.reject(1)`.

**Comment by aab-odoo** [comment]
@xmo-odoo `finally` is not ES6...

**Comment by xmo-odoo** [comment]
> @xmo-odoo `finally` is not ES6...

Oh yeah… well crap.

**Comment by aab-odoo** [comment]
Indeed :(

**Comment by keshrath** [comment]
I hope it's ok if I ask this question directly in PR, but I didn't want to open a separate ticket for it. What does that mean in terms of codestyle and ES version?
Should be used in the future let/const and Arrow Functions. In the comment here I read that you don't use the finally method because it's not ES6. So the jump is ES5 to ES6? I've also seen async and await functions in the code wouldn't that require ES8?

**Comment by aab-odoo** [comment]
We must still support IE11 for v13, so we keep coding in ES5 for now (except for the Promise, for which we use a polyfill). This doesn't apply to tests, for which we can use ES8 features.

**Pull request merged**